### ISSUE TITLE
Internal ID links unable to work due to partial page rendering

### ISSUE THREAD

**Issue by 5e-Cleric** [naturalcrit/homebrewery#1923]
Partial page rendering blocks ID selector links, as you can check in [this document](https://homebrewery.naturalcrit.com/share/15P8uvJD9xYFWDb_wPkNwiGJCSlZKykbU7Xdi03qd1klN), if you delete one of the pages, the partial page rendering deactivates and the link is able to locate its destination.

However, page number (which is also an id) does work perfectly on both situations.

Linking to IDs is very useful when working on long brews that keep growing and it would be nice if this got fixed.

**Comment by ericscheid** [comment]
> However, page number (which is also an id) does work perfectly on both situations.

The non-rendered pages are still inserted into the DOM as an empty placeholder with the `id`, like this:
```
<div class="phb page" id="p106">
  <i class="fas fa-spinner fa-spin"></i>
</div>
```

When those placeholders are scrolled into view, then they are rendered.

----

Partial page rendering kicks in with larger brews (as you've noticed) — this is a performance optimisation as re-rendering a great many pages is quite slow. (The conversion of _md → html_ is the main bottleneck, not the browser coping with a huge html body).

However, we could consider an alternative strategy: only re-render the pages that were edited, leave the rest untouched. (but also re-render the page above and below the edited page, as the edit might actually be an addition/removal of `\page`)

The simpler form of this strategy would require rendering the entire brew at the time of loading. This could be quite onerous and users might think the newly-loaded page has crashed somehow. So we'd eventually need a better progress indicator than a spinning gif.

The more complicated form of this strategy would start with partial-page-rendering, but in each of those rendered pages we check if there are any anchor links and then do a search for pages that contain that anchor and render them too. This is getting messy, and is indistinguishable from the "dumbly render everything" variant if there's a TOC with links to all pages, so we're back to the "has it crashed?" scenario again. TOC links are usually to `id="p<number>"`, and they work regardless (as you've noted), so they could be skipped in this algo (deferring rendering when they are scrolled into view). Complicated.

----

Side note: IIRC, react has the built-in ability to only update those parts of the DOM that have changed, so that can be leveraged here too. (For example, don't change pages that have scrolled outside of the _view±3pages_ back to placeholder divs).

----

TL;DR: this is achievable, but is not a trivial thing.
